cores/esp8266/core_esp8266_main.cpp

```diff
@@ -243,18 +243,26 @@ extern "C" void ICACHE_RAM_ATTR app_entry (void)
     return app_entry_custom();
 }
 
+extern "C" void preinit (void) __attribute__((weak));
+extern "C" void preinit (void)
+{
+    /* do nothing by default */
+}
+
 extern "C" void user_init(void) {
     struct rst_info *rtc_info_ptr = system_get_rst_info();
     memcpy((void *) &resetInfo, (void *) rtc_info_ptr, sizeof(resetInfo));
 
     uart_div_modify(0, UART_CLK_FREQ / (115200));
 
-    init();
+    init(); // in core_esp8266_wiring.c
 
     initVariant();
 
     cont_init(g_pcont);
 
+    preinit(); // user redefinable
```

Comment:
To reduce confusion, please add a comment that this function refers to Before C++ Dynamic Init phase, and is not related to the init() above, which is about initializing hw registers and sdk timers.
Refs: 
https://en.cppreference.com/w/cpp/language/initialization
https://stackoverflow.com/questions/5945897/what-is-dynamic-initialization-of-object-in-c